File: src/ai/flows/score-and-parse-resume.ts
'use server';

/**
 * @fileOverview An AI agent that parses a resume and scores it against a job description.
 *
 * - scoreAndParseResume - A function that handles the resume parsing and scoring process.
 * - ScoreAndParseResumeInput - The input type for the scoreAndParseResume function.
 * - ScoreAndParseResumeOutput - The return type for the scoreAndParseResume function.
 */

import {ai} from '@/ai/genkit';
import {z} from 'genkit';

const ScoreAndParseResumeInputSchema = z.object({
  jobDescription: z.string().describe('The job description for the role.'),
  resumeDataUri: z
    .string()
    .describe(
      "A resume file (PDF, DOCX) or an image of a resume, as a data URI that must include a MIME type and use Base64 encoding. Expected format: 'data:<mimetype>;base64,<encoded_data>'."
    ),
});
export type ScoreAndParseResumeInput = z.infer<typeof ScoreAndParseResumeInputSchema>;

const WorkExperienceSchema = z.object({
  company: z.string().describe('The name of the company.'),
  title: z.string().describe('The job title or role.'),
  dates: z.string().describe('The dates of employment.'),
});

const EducationSchema = z.object({
  institution: z.string().describe('The name of the educational institution.'),
  degree: z.string().describe('The degree or qualification obtained.'),
  year: z.string().describe('The year of graduation or completion.'),
});

const ScoreAndParseResumeOutputSchema = z.object({
  score: z
    .number()
    .describe('The score (0-100) of the resume based on the job description.'),
  justification: z
    .string()
    .describe('The justification for the assigned score.'),
  parsedData: z.object({
    name: z.string().describe("The candidate's full name."),
    email: z.string().describe("The candidate's email address."),
    phone: z.string().describe("The candidate's phone number."),
    links: z.array(z.string()).describe("Array of URLs for social profiles like LinkedIn, GitHub, etc."),
    skills: z.array(z.string()).describe('An array of skills extracted from the resume.'),
    workExperience: z.array(WorkExperienceSchema).describe('An array of work experience objects.'),
    education: z.array(EducationSchema).describe('An array of education objects.'),
  }),
});
export type ScoreAndParseResumeOutput = z.infer<typeof ScoreAndParseResumeOutputSchema>;

export async function scoreAndParseResume(input: ScoreAndParseResumeInput): Promise<ScoreAndParseResumeOutput> {
  return scoreAndParseResumeFlow(input);
}

const prompt = ai.definePrompt({
  name: 'scoreAndParseResumePrompt',
  input: {schema: ScoreAndParseResumeInputSchema},
  output: {schema: ScoreAndParseResumeOutputSchema},
  prompt: `You are an expert HR recruiter with experience in parsing resumes and matching candidates to job descriptions.

You will be provided with a job description and a resume (either as text from a document or from an image using OCR). Your tasks are:
1.  Parse the resume to extract structured information: candidate's name, email, phone, social links (LinkedIn, GitHub, etc.), skills, work experience (company, title, dates), and education (institution, degree, year).
2.  Score the resume from 0-100 based on how well the candidate's skills and experience match the provided job description.
3.  Provide a concise justification for the score.

Job Description: {{{jobDescription}}}

Resume Content:
{{media url=resumeDataUri}}

Please return the extracted data, score, and justification in the specified JSON format.
`,
});

const scoreAndParseResumeFlow = ai.defineFlow(
  {
    name: 'scoreAndParseResumeFlow',
    inputSchema: ScoreAndParseResumeInputSchema,
    outputSchema: ScoreAndParseResumeOutputSchema,
  },
  async input => {
    const {output} = await prompt(input);
    return output!;
  }
);

File: python-ai-gemini/src/flows/score_and_parse_resume.py
import base64
import mimetypes
from typing import Optional
from pydantic import BaseModel, Field
from ..client import GeminiClient
from ..schemas import ParsedResume, WorkExperience, Education, Project

class ScoreAndParseResumeInput(BaseModel):
    """Input for resume scoring and parsing."""
    job_description: str = Field(description="The job description for the role")
    resume_data_uri: str = Field(
        description="A resume file (PDF, DOCX) or an image of a resume, as a data URI"
    )

class ScoreAndParseResumeOutput(BaseModel):
    """Output from resume scoring and parsing."""
    score: int = Field(ge=0, le=100, description="The score (0-100) of the resume")
    justification: str = Field(description="A concise justification for the assigned score")
    parsed_data: ParsedResume = Field(description="The parsed resume data")

class ScoreAndParseResumeFlow:
    """AI agent that parses resumes and scores them against job descriptions."""
    
    def __init__(self, client: GeminiClient):
        self.client = client
    
    def _process_data_uri(self, data_uri: str) -> tuple[str, bytes]:
        """Process data URI to extract MIME type and data."""
        if not data_uri.startswith("data:"):
            raise ValueError("Invalid data URI format")
        
        header, data = data_uri.split(",", 1)
        mime_type = header.split(";")[0].split(":")[1]
        
        # Decode base64 data
        try:
            decoded_data = base64.b64decode(data)
        except Exception as e:
            raise ValueError(f"Failed to decode base64 data: {e}")
        
        return mime_type, decoded_data
    
    async def execute(self, input_data: ScoreAndParseResumeInput) -> ScoreAndParseResumeOutput:
        """Execute the resume scoring and parsing flow."""
        
        # Process the data URI
        mime_type, file_data = self._process_data_uri(input_data.resume_data_uri)
        
        prompt = f"""You are an expert HR recruiter with experience in parsing resumes and matching candidates to job descriptions.

You will be provided with a job description and a resume. Your tasks are:
1. Parse the resume to extract structured information. Be as accurate as possible. Extract all fields defined in the output schema.
2. Score the resume from 0 to 100 based on how well the candidate's skills and experience match the provided job description.
3. Provide a concise justification for the score. Explain the reasoning behind your score, noting how the candidate aligns with the requirements.

Job Description:
```
{input_data.job_description}
```

Please return the extracted data, score, and justification as JSON with the following structure:
{{
    "score": <integer 0-100>,
    "justification": "<string>",
    "parsed_data": {{
        "name": "<string>",
        "email": "<string>",
        "phone": "<string>",
        "links": ["<string>"],
        "summary": "<string>",
        "skills": ["<string>"],
        "work_experience": [
            {{
                "company": "<string>",
                "title": "<string>",
                "dates": "<string>"
            }}
        ],
        "education": [
            {{
                "institution": "<string>",
                "degree": "<string>",
                "year": "<string>"
            }}
        ],
        "projects": [
            {{
                "name": "<string>",
                "description": "<string>",
                "url": "<string>"
            }}
        ],
        "certifications": ["<string>"],
        "languages": ["<string>"],
        "hobbies": ["<string>"]
    }}
}}

If a field is not present in the resume, return an empty string or empty array for it."""
        
        response_schema = {
            "type": "object",
            "properties": {
                "score": {"type": "integer", "minimum": 0, "maximum": 100},
                "justification": {"type": "string"},
                "parsed_data": {
                    "type": "object",
                    "properties": {
                        "name": {"type": "string"},
                        "email": {"type": "string"},
                        "phone": {"type": "string"},
                        "links": {"type": "array", "items": {"type": "string"}},
                        "summary": {"type": "string"},
                        "skills": {"type": "array", "items": {"type": "string"}},
                        "work_experience": {
                            "type": "array",
                            "items": {
                                "type": "object",
                                "properties": {
                                    "company": {"type": "string"},
                                    "title": {"type": "string"},
                                    "dates": {"type": "string"}
                                },
                                "required": ["company", "title", "dates"]
                            }
                        },
                        "education": {
                            "type": "array",
                            "items": {
                                "type": "object",
                                "properties": {
                                    "institution": {"type": "string"},
                                    "degree": {"type": "string"},
                                    "year": {"type": "string"}
                                },
                                "required": ["institution", "degree", "year"]
                            }
                        },
                        "projects": {
                            "type": "array",
                            "items": {
                                "type": "object",
                                "properties": {
                                    "name": {"type": "string"},
                                    "description": {"type": "string"},
                                    "url": {"type": "string"}
                                },
                                "required": ["name", "description"]
                            }
                        },
                        "certifications": {"type": "array", "items": {"type": "string"}},
                        "languages": {"type": "array", "items": {"type": "string"}},
                        "hobbies": {"type": "array", "items": {"type": "string"}}
                    },
                    "required": ["name", "skills", "work_experience", "education", "certifications", "languages", "links"]
                }
            },
            "required": ["score", "justification", "parsed_data"]
        }
        
        # For file processing, we'll use text-based approach
        # In a real implementation, you might use Gemini's multimodal capabilities
        result = await self.client.generate_structured(
            prompt=prompt,
            response_schema=response_schema
        )
        
        # Convert to Pydantic models
        parsed_data = ParsedResume(
            name=result["parsed_data"]["name"],
            email=result["parsed_data"].get("email"),
            phone=result["parsed_data"].get("phone"),
            links=result["parsed_data"]["links"],
            summary=result["parsed_data"].get("summary"),
            skills=result["parsed_data"]["skills"],
            work_experience=[
                WorkExperience(**exp) 
                for exp in result["parsed_data"]["work_experience"]
            ],
            education=[
                Education(**edu) 
                for edu in result["parsed_data"]["education"]
            ],
            projects=[
                Project(**proj) 
                for proj in result["parsed_data"].get("projects", [])
            ] if result["parsed_data"].get("projects") else None,
            certifications=result["parsed_data"]["certifications"],
            languages=result["parsed_data"]["languages"],
            hobbies=result["parsed_data"].get("hobbies")
        )
        
        return ScoreAndParseResumeOutput(
            score=result["score"],
            justification=result["justification"],
            parsed_data=parsed_data
        )

async def score_and_parse_resume(job_description: str, resume_data_uri: str) -> ScoreAndParseResumeOutput:
    """Convenience function for scoring and parsing resumes."""
    client = GeminiClient()
    flow = ScoreAndParseResumeFlow(client)
    return await flow.execute(ScoreAndParseResumeInput(
        job_description=job_description,
        resume_data_uri=resume_data_uri
    ))
